pkg/apis/triggers/v1beta1/event_listener_defaults.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2019 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"context"
+
+	"github.com/tektoncd/triggers/pkg/apis/triggers/contexts"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/ptr"
+)
+
+// SetDefaults sets the defaults on the object.
+func (el *EventListener) SetDefaults(ctx context.Context) {
+	if contexts.IsUpgradeViaDefaulting(ctx) {
+		// set defaults
+		if el.Spec.Resources.KubernetesResource != nil {
+			if el.Spec.Resources.KubernetesResource.Replicas != nil && *el.Spec.Resources.KubernetesResource.Replicas == 0 {
+				*el.Spec.Resources.KubernetesResource.Replicas = 1
+			}
+		}
+
+		for i, t := range el.Spec.Triggers {
+			triggerSpecBindingArray(el.Spec.Triggers[i].Bindings).defaultBindings()
+			for _, ti := range t.Interceptors {
+				ti.defaultInterceptorKind()
+				if err := ti.updateCoreInterceptors(); err != nil {
+					// The err only happens due to malformed JSON and should never really happen
+					// We can't return an error here, so print out the error
+					logger := logging.FromContext(ctx)
+					logger.Errorf("failed to setDefaults for trigger: %s; err: %s", t.Name, err)
+				}
+			}
+		}
+		el.Spec.updatePodTemplate()
+		// To be removed in a later release #1020
+		el.Spec.updateReplicas()
+	}
+}
+
+func (spec *EventListenerSpec) updatePodTemplate() {
+	if spec.DeprecatedPodTemplate != nil {
```

Comment:
That particular field is going away with https://github.com/tektoncd/triggers/issues/1105 though generally it depends - we'd have to keep supporting fields in v1beta1 that are a part of v1alpha1 due to how kube api machinery works (v1beta1 is the stored version so all fields must be supported). We'll use the validation webhook to return errors for fields that should not be supported in v1beta1 version.

Reply:
There's also other deprecated fields for interceptors, etc. My worry is if we introduce these fields, if we're strictly following the deprecation policy this effectively resets the clock for removal since we're "introducing a new field" even though it's deprecated from the start.

If I'm understanding https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#on-compatibility correctly, we don't need to have 1:1 parity with every field in the stored version - we just need to make sure users can go from v1alpha1 to v1beta1 and back without data loss.
The same way [we already mutate `DeprecatedPodTemplate` -> `spec.Resources.KubernetesResource`](https://github.com/tektoncd/triggers/blob/7c72387dd1144c648fecae0746174fab970c6f4a/pkg/apis/triggers/v1alpha1/event_listener_defaults.go#L55-L71), I don't think we need to support the deprecated fields in v1beta1, since we can safely convert back to v1alpha1 by using `KubernetesResource`.  This is analogous to the k8s example where v6 holds (Param, Params), where as the stored v7 version only stores Params and does not preserve the param field.

Reply:
I think that's a valid concern. I think there were some issues with JSON marshalling without having the fields set but let me verify. I think given the conversions we are doing in the defaulting webhook, we should not need an extra conversion webhook

Reply:
Ok, I think this will actually work!

Reply:
I assume it works because when we upgrade from old version to new version there won't be API update so if user want they will create new resources with `v1beta1` am i right ?